### ISSUE TITLE
Embedded languages

### ISSUE THREAD

**Issue by ivan-kleshnin** [jneen/parsimmon#251]
**Question: how to address `line:column` mismatches for embedded languages?**

For example, I have the following file:

```
---
foo: bar
---

Baz
```

in FrontMatter format and the following parser, working with `fm.body`:

```js
let P = require("parsimmon")

let s = `
Baz
`

let lang = P.createLanguage({
  Baz: function (L) {
    return P.string("Baz+").trim(P.optWhitespace)
  },
})

console.log(lang.Baz.tryParse(s))
```

The error says:

```
Error: expected 'Baz+' at line 2 column 1, got '...Baz
```

While the real line, as the **user** perceives it is `5`. The solution of catching and parsing exception messages with `.replace(/line (\d+)/, "...")` is obviously a last resort hack.

**Comment by wavebeem** [comment]
`tryParse` is a convenience method for common use cases. I would recommend using `parse` instead which returns a data object describing what happened: <https://github.com/jneen/parsimmon/blob/master/API.md#parserparseinput>

The return value of `parse` should be useful in constructing a custom error message for your use case:

```js
({
  status: false,
  index: {offset: 0, line: 1, column: 1},
  expected: ['EOF', 'whitespace']
})
```

**Comment by wavebeem** [comment]
Alternatively you might want to compose your front matter parser and your body parset together, and only call .tryParsr once so it tracks the positions correctly

**Comment by wavebeem** [comment]
Like instead of doing:

* Parse front matter, get raw body text
* Parse raw body text with parser based on file extension

You could do like:

```js
let MarkdownParser = P.createLanguage({ ... });
let FrontmatterParser = P.createLanguage({ ... });
let EnhancedMarkdownParser = P.seq(FrontmatterParser.main, MarkdownParser.main);

EnhancedMarkdownParser.tryParse(fileTextContent);
```

or something like that?

**Comment by ivan-kleshnin** [comment]
Yeah, I thought of this version but wanted to avoid writing a YAML parser.

Now I see I can create just-an-FM parser like: 

```js
P.seq(
  P.string("---")).skip(P.nl),
  P.any.notFollowedBy(P.string("---")).many().tie(),
  P.nl.then(P.string("---")),
)
```

combined with body parser as you said.

Then parse YAML with a 3-rd party lib after that, on a separate step. That would raise the original problem again, now for the YAML, which would be 1 line shifted from it's initial position. But than I can fake it by adding one empty line to the YAML parser before applying. Problem solved 😃
